Remove some indirections in Global and Typer

retronym opened the pull request:
This reduces the number of Java stack frames between the bottom
of the stack and the start of typer, and, more importantly,
at each level of recursion in the AST or in symbol completion.

The intent is to make the stacks easier to visually scan in
profilers and other tools that display them. I'm not expecting
that performance will improve, the JVM probably does a decent
jobs and inlining these chunks of the stack.

retronym committed:
Remove some indirections in Global and Typer

This reduces the number of Java stack frames between the bottom
of the stack and the start of typer, and, more importantly,
at each level of recursion in the AST or in symbol completion.

The intent is to make the stacks easier to visually scan in
profilers and other tools that display them. I'm not expecting
that performance will improve, the JVM probably does a decent
jobs and inlining these chunks of the stack.

retronym commented:
Reviving #5924

lrytz commented on src/compiler/scala/tools/nsc/typechecker/TypersTracking.scala:
```diff
@@ -138,6 +138,16 @@ trait TypersTracking {
+      showPush(tree, mode, pt, context)
+      true
+    }
+    def afterNextTyped(tree: Tree, typedTree: Tree): Unit = {
```
not invoked?

lrytz reviewed (commented):
it seems the higher-order methods in TypersTracking are no longer used, should they be removed?

retronym committed:
Remove now unused methods from TypersTracking

retronym commented:
@lrytz I've removed unused methods. The failure to call `afterNextTyped` was because that action is now split up into two lines of code, one [inside the try](https://github.com/scala/scala/blob/3d69134a50cc1e010e135377c54a7375db13ddb9/src/compiler/scala/tools/nsc/typechecker/Typers.scala#L5659) an one [in the finally](https://github.com/scala/scala/blob/3d69134a50cc1e010e135377c54a7375db13ddb9/src/compiler/scala/tools/nsc/typechecker/Typers.scala#L5678).

lrytz reviewed (approved):
Thanks!
